openshift/machine-config-operator
# MCO-501: Infer OS version from image labels

**- What I did**

- I've added code that can infer what version the new OS is from the image labels. This could be useful in future cases where we might need to perform some type of pre-work before applying a new OS image.
- Like with the node OS detection code, I've wired it into an e2e test so that we are constantly testing this code path.
- I've also modified the MCD to use an interface for OS detection, allowing us to substitute a fake OS implementation for testing. This will allow us to easily reach OS-dependent code-paths without much overhead.

*NOTE:* Except for the e2e test, this code does not actually query the image registry which contains the new node OS container. It only parses the results from it.

**- How to verify it**

- Run the unit tests to verify that the new label inference code works correctly.
- Run the e2e tests and verify that a line resembling: `node_os_test.go:38: Found OS image "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:329a8968765c2eca37d8cbd95ecab0400b5252a680eea9d279f41d7a8b4fdb93" matching tag "rhel-coreos-8" in OCP / OKD release image "registry.ci.openshift.org/ocp/release@sha256:21a7209435eb0461c1a931e0245b9e77ea7fe9ae266acca90e53b93cb2dbc4ea". Will compare to what is on cluster nodes` is contained therein.

**- Description for the changelog**
Enables inferring what the next node OS will be before upgrading
